core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -1425,6 +1460,7 @@ private boolean hasChanges() {
           || (discardChanges && !changes.isEmpty())
           || metadataLocation != null
           || suppressHistoricalSnapshots
+          || hasRemovedSpecs
```

Comment:
Hm I'm trying to understand why we need this special flag, is this a way so that we avoid having to add the metadata update type for REST (since then that would ultimately just be in the changes list)?

Reply:
This is not the right way to update `hasChanges`. Instead, this needs to add a change to `changes`. That way it is sent to REST services to modify the table in the REST commit path.

Reply:
> Yes, as discussed earlier, I don't think we should expose removePartitionSpec directly to the REST API as there's no easy way to ensure that the spec to remove is indeed not used.

We don't currently have operations that can't be supported by the REST protocol, so this is an area where we should be careful. I agree that we want to ensure that there are no new references to specs that are being removed, but I'm skeptical that there is no way to do that. There are also problems with not sending this because it would be a silent no-op when sending changes to REST catalogs.

It's unlikely that a new snapshot would be written with a spec that is being removed because this already validates that the default spec is not being removed. A conflict here would require that a concurrent writer is using a spec other than the default or has changed the default spec. There's already a validation for the second case, `assert-default-spec-id`. For the first case, this could require that no branch states have changed using `assert-ref-snapshot-id`.

Reply:
> The RemoveUnusedSpecs will succeed without actually removing unused specs for REST catalog.
@advancedxy This indicates there's a problem with the current implementation then imo. If there was a way to avoid the REST spec change I think it would've been OK as long as we can guarantee failure on the client side until the support was added for the metadata update type. But I think that's unavoidable, and I agree with @rdblue that we should probably just add the metadata update type. 

I also am reasonably confident that a REST catalog can safely handle this RemovePartitionSpec update.